File: apps/Frontend Plane/velionv3/src/features/onboarding/lib/api/billing.ts
import { requestJson } from '@/shared/api/http'
import { withDevActor } from '@/features/onboarding/lib/api/actor'
import type { ActionActor } from '@/features/onboarding/lib/api/contracts'
import type { CheckoutSession, CheckoutStatus } from '@/features/billing/lib/api'

export type { CheckoutSession, CheckoutStatus } from '@/features/billing/lib/api'

export async function setPlan(input: {
  actor: ActionActor
  orgId: string
  plan: string
  onboarding?: Record<string, unknown>
}): Promise<{ id: string; plan: string }> {
  return requestJson('/api/v1/onboarding/actions/set-plan', {
    method: 'POST',
    body: JSON.stringify(withDevActor({
      orgId: input.orgId,
      plan: input.plan,
      reason: 'onboarding',
      onboarding: input.onboarding,
    }, input.actor)),
  })
}

export async function startCheckout(input: {
  actor: ActionActor
  orgId: string
  plan: string
  successUrl: string
  cancelUrl: string
}): Promise<CheckoutSession> {
  return requestJson('/api/v1/onboarding/actions/start-checkout', {
    method: 'POST',
    body: JSON.stringify(withDevActor({
      orgId: input.orgId,
      plan: input.plan,
      successUrl: input.successUrl,
      cancelUrl: input.cancelUrl,
    }, input.actor)),
  })
}

export async function confirmCheckout(input: {
  actor: ActionActor
  orgId: string
  plan: string
  paymentId?: string
  clientSecret?: string
}): Promise<CheckoutStatus> {
  return requestJson('/api/v1/onboarding/actions/confirm-checkout', {
    method: 'POST',
    body: JSON.stringify(withDevActor({
      orgId: input.orgId,
      plan: input.plan,
      paymentId: input.paymentId,
      clientSecret: input.clientSecret,
    }, input.actor)),
  })
}

File: apps/Frontend Plane/velionv3/apps/gateway/src/onboarding/actions/billing.rs
use axum::{extract::State, http::HeaderMap, response::IntoResponse, Json};
use reqwest::Method;
use serde_json::json;

use crate::{
    auth::actor_from_request,
    config::AppState,
    contracts::{ConfirmCheckoutRequest, SetPlanRequest, StartCheckoutRequest},
    upstream::proxy_json,
};

pub(crate) async fn set_plan(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(input): Json<SetPlanRequest>,
) -> impl IntoResponse {
    let actor = actor_from_request(
        input.actor.as_ref(),
        Some(&headers),
        state.allow_dev_actor_headers,
    );
    let body = json!({
        "plan": input.plan,
        "reason": input.reason.unwrap_or_else(|| "onboarding".into()),
        "onboarding": input.onboarding,
    });

    proxy_json(
        &state,
        Method::POST,
        &format!(
            "{}/orgs/{}/plan",
            state.org_core_url,
            urlencoding::encode(input.org_id.trim())
        ),
        Some(body),
        Some(input.org_id.trim()),
        Some(&actor),
        None,
    )
    .await
}

pub(crate) async fn start_checkout(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(input): Json<StartCheckoutRequest>,
) -> impl IntoResponse {
    let actor = actor_from_request(
        input.actor.as_ref(),
        Some(&headers),
        state.allow_dev_actor_headers,
    );
    let body = json!({
        "plan": input.plan,
        "success_url": input.success_url,
        "cancel_url": input.cancel_url,
    });

    proxy_json(
        &state,
        Method::POST,
        &format!(
            "{}/api/v1/billing/orgs/{}/checkout-session",
            state.billing_core_url,
            urlencoding::encode(input.org_id.trim())
        ),
        Some(body),
        Some(input.org_id.trim()),
        Some(&actor),
        None,
    )
    .await
}

pub(crate) async fn confirm_checkout(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(input): Json<ConfirmCheckoutRequest>,
) -> impl IntoResponse {
    let actor = actor_from_request(
        input.actor.as_ref(),
        Some(&headers),
        state.allow_dev_actor_headers,
    );
    let body = json!({
        "plan": input.plan,
        "payment_id": input.payment_id,
        "client_secret": input.client_secret,
    });

    proxy_json(
        &state,
        Method::POST,
        &format!(
            "{}/api/v1/billing/orgs/{}/checkout-session/confirm",
            state.billing_core_url,
            urlencoding::encode(input.org_id.trim())
        ),
        Some(body),
        Some(input.org_id.trim()),
        Some(&actor),
        None,
    )
    .await
}
